presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchangeSource.java

```diff
@@ -0,0 +1,216 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.exchange;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class LocalExchangeSource
+{
+    private static final SettableFuture<?> NOT_EMPTY;
+    static {
+        NOT_EMPTY = SettableFuture.create();
+        NOT_EMPTY.set(null);
+    }
+
+    private final List<Type> types;
+    private final Consumer<LocalExchangeSource> onFinish;
+
+    private final BlockingQueue<PageReference> buffer = new LinkedBlockingDeque<>();
+    private final AtomicLong bufferedBytes = new AtomicLong();
+
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
+    private SettableFuture<?> notEmptyFuture = NOT_EMPTY;
+
+    @GuardedBy("lock")
+    private boolean finishing;
+
+    public LocalExchangeSource(List<? extends Type> types, Consumer<LocalExchangeSource> onFinish)
+    {
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.onFinish = requireNonNull(onFinish, "onFinish is null");
+    }
+
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    public LocalExchangeBufferInfo getBufferInfo()
+    {
+        // This must be lock free to assure task info creation is fast
+        // Note: the stats my be internally inconsistent
+        return new LocalExchangeBufferInfo(bufferedBytes.get(), buffer.size());
+    }
+
+    void addPage(PageReference pageReference)
+    {
+        checkNotHoldsLock();
+
+        boolean added = false;
+        SettableFuture<?> notEmptyFuture;
+        synchronized (lock) {
+            // ignore pages after finish
+            if (!finishing) {
+                // buffered bytes must be updated before adding to the buffer to assure
+                // the count does not go negative
+                bufferedBytes.addAndGet(pageReference.getSizeInBytes());
+                buffer.add(pageReference);
+                added = true;
+            }
+
+            // we just added a page (or we are finishing) so we are not empty
+            notEmptyFuture = this.notEmptyFuture;
+            this.notEmptyFuture = NOT_EMPTY;
+        }
+
+        if (!added) {
+            // dereference the page outside of lock
+            pageReference.removePage();
```

Comment:
This looks wrong. The way I understand it, the page need to be removed from the PageReference anyways.